### ISSUE TITLE
Allow RenderDoc to launch and capture native Windows Arm64 apps

### ISSUE THREAD

**Issue by ssegreto** [baldurk/renderdoc#2298]
## Description

I am porting a Windows DX12 project to the Surface Pro X and found that RenderDoc can't launch my app when it's built as native Windows Arm64, because the profiler shim isn't built for native Arm64.
Could this support be added to RenderDoc please?

A possible solution is to follow the same approach as the x86 profiler shim and build the same RenderDoc pieces for Arm64.
That will require upgrading RenderDoc's Windows build environment to VS2017+ instead of VS2015 and some minor porting to Arm64. (May not be too many changes from quick review..)
If the x64 profiler shim can't be injected, the tool could try to inject the Arm64 version instead or maybe the user could hint to the launcher with UI options to use: Arm64 instead of x64.

## Environment

* RenderDoc version: v1.14
* Operating System: Windows 10 Version 21390 MP (8 procs) Free ARM 64-bit (AArch64)
* Graphics API: DirectX 12

**Comment by baldurk** [comment]
Sorry, but I have no plans to support ARM builds on desktop.

**Comment by ssegreto** [comment]
Can I submit a PR to provide this support for RenderDoc?

**Comment by baldurk** [comment]
In general I don't mind accepting PRs for unsupported platforms as long as it doesn't affect or break the supported platforms, but if it's true that you'd need to change the VS projects to target something newer than VS2015 then that wouldn't be something I'd accept.

**Comment by ssegreto** [comment]
Unfortunately it seems Microsoft has no plans to backport Arm64 support to VS2015, so it would require updating RenderDoc to use VS2017. Arm32 can be supported by VS2015, but folks porting applications to ARM Windows are usually porting from x64 to Arm64 directly. Is there any timeline of when RDC would move to VS2017? [edited at author's request]

**Comment by baldurk** [comment]
I don't have any timeline for changing VS version. It's primarily decided by the project's library dependencies, that's what forced the change to 2015 previously. If the dependencies require a newer compiler version and it's completely infeasible to find a way to remain on 2015 then I'll have no choice, but that may be a long time from now.

**Comment by ssegreto** [comment]
Thanks for the replies. Right now the UI dependencies of RenderDoc don't have immediate Windows Arm64 versions (qtWidgets and Google Breakpad), so my plan was to continue to have those portions of RDC as x64 (Microsoft has an emulation that works well for those portions). However the binary that is injected into the user's graphical process must match the hosting processes architecture, which is why I need an Arm64 RDC profiler shim. Would you be amenable to some sort of hybrid build setup that used VS2017 just to build Arm64 versions of renderdoccmd.exe, dbghelp.dll, renderdoc.dll, renderdocshim64.dll, symsrv.dll, renderdoc.json, symsrv.yes? The remainder of the RDC solution would remain identical and continue to use VS2015.

**Comment by baldurk** [comment]
I thought that was probably what you meant, but unless I misunderstand that would still require the renderdoc solution and core projects to be converted to VS2017 or newer, which is what I don't want to do.

**Comment by ssegreto** [comment]
There could be two projects and solutions, the original vs2015 set and a smaller subset that is just for Arm64 and has VS2017 "copies" of the needed projects in a new solution, then the scripts (buildscripts/scripts/compile_win32.sh and buildscripts/scripts/make_package_win32.sh) could call out to a new set of Arm64 scripts that use VS2017 and a final Windows package would have an x86 and an Arm64 additional architecture folders.

**Comment by baldurk** [comment]
No I wouldn't want to do that, that would be too intrusive for an unsupported platform and increases the maintenance burden for any build changes.

**Comment by ssegreto** [comment]
Thanks for hearing out my case. One final question? How do you ingest the 3rd party dependencies for RenderDoc like breakpad, exception handling and compressonator, etc. The reason I ask is while experimenting with this feature request I noticed I had to make minor preproc changes to support Win Arm64 (e.g. to avoid including Intel SSE intrinsic headers or to support Arm64 context records, etc). Do you just have these static copies as part of RenderDoc or do you periodically refresh them from the source, etc?

**Comment by baldurk** [comment]
No problem. I'm sorry I can't add support for this but it doesn't make sense from the perspective of priorities and maintenance cost.

All the dependencies are copied into the repository here to ensure that anyone can build on windows without needing extra setup. They aren't updated on a set schedule, just when it's necessary either to fix a bug or add a desired feature. Most of the dependencies have a relatively specific use-case so as long as they continue to work there's not a high amount of churn.

**Comment by ssegreto** [comment]
So realistically Win Arm64 support would need to be added into the aforementioned dependencies rather than edited into the Renderdoc copies of them, otherwise the next dep update could break things (assuming RDC ever did decide to take Arm64 support)?

**Comment by baldurk** [comment]
Generally when I update dependencies I preserve any local changes that have been made, so it wouldn't be a hard requirement necessarily. For more complex dependencies like breakpad though you could simply compile without crash-handling on ARM, since that is not relevant unless a stable release is ever made (non-stable/development/bespoke builds do not submit crashes so that the reports aren't polluted with buggy or custom builds).

**Comment by Biswa96** [comment]
If possible a separate VS solution and project files for VS2017 or VS2019 can be saved in this project.

**Comment by ssegreto** [comment]
The reality is that Pix for Windows supports more scenarios than RenderDoc when it comes to Windows on Arm. Would love to help change that.

**Comment by baldurk** [comment]
I'm not sure what else there is to say. As above I have no plan to support this at all, and that while I don't mind accepting PRs for fixes/handling of unsupported platforms, that can't come with either a maintenance cost or potential bugs for the supported platforms so the above constraints about not breaking or splitting the build system remain.

**Comment by Biswa96** [comment]
There is only two issues. zstd and Qt5 dependencies don't have ARM64 support. Other than that everything compiles well (with some little tweaking).

**Comment by Biswa96** [comment]
Little question. Do you update/fix projects in renderdoc/3rdparty folder manually or copy-paste directly from upstream without any modifying any code?

**Comment by baldurk** [comment]
External projects aren't automatically or regularly updated, I update them as and when needed to get a fix or improvement from upstream that I want. Any changes I've made locally I preserve either by manually applying the diff onto the latest version or by merging, whichever makes most sense.

I don't want this issue to become a permanent discussion as that's not how the issue tracker is used. The original request has been addressed as I have no plans to support ARM on desktop, so I'm locking this as the issue has been resolved.
